### PR TITLE
feat(cognito): user pool: adds custom sender (Email/SMS) lambda triggers

### PR DESCRIPTION
**Pull request by thchia**
### Motivation

I would like to use the Custom Sender triggers in my CDK app. They are supported by raw CF (despite the docs claiming otherwise).

I do not wish to use the CfnUserPool construct as I am unable to pass the pool as a reference to other constructs (e.g. API Gateway authorisers/User Pool App Clients/etc.).

I do not wish to follow the method recommended by the docs either (using the CLI to update the pool post-deployment), as it is hard to automate (requiring me to fetch the ARNs of my functions and KMS Key, not to mention it overrides other properties of the User Pool), and leaks my stack configuration out from CDK and into some post-deploy script.

Please find this PR in support of adding the triggers to CDK.

### Description

Adds support for 2 extra lambda triggers (`CustomEmailSender` and `CustomSMSSender`). These triggers have a different format, requiring both function ARN and version (only supported value is `V1_0`).

Additionally, specifying either of these requires a `KMSKeyId` be specified as well, which appears as a property of the `LambdaTriggers` property in CF.

Additionally, specifying the `CustomSMSSender` requires the `smsRole` be specified (allowing `sns:publish`).

Neither of these requirements are enforced in the code, but CF will throw helpful (IMO) errors if they are not satisfied.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-cognito/lib/user-pool.ts [hunk 1]
```diff
@@ -138,10 +138,31 @@ export interface UserPoolTriggers {
...
+   * @see https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-custom-sender-triggers.html
+   * @default - no key ID configured
+   */
+  readonly kmsKeyId?: string
+
   /**
    * Index signature
    */
-  [trigger: string]: lambda.IFunction | undefined;
+  [trigger: string]: lambda.IFunction | string | undefined;
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool.ts:165]: union types such as these are not supported by jsii.

In Java, they will get incorrectly translated into `java.lang.Object`.

Move the kms key to the top-level `UserPoolProps`.
Call it `customSenderKmsKey` with type `IKey` from the `@aws-cdk/aws-kms` module?

File: packages/@aws-cdk/aws-cognito/test/user-pool.test.ts [hunk 2]
```diff
@@ -331,6 +331,65 @@ describe('User Pool', () => {
     });
   });
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/test/user-pool.test.ts:333]: Can you add a new integration test of this nature - https://github.com/aws/aws-cdk/blob/c62377e14caae677deb7e4eae692eaccb2020c67/packages/%40aws-cdk/aws-cognito/test/integ.user-pool-idp.apple.ts#L4-L12 - with verification steps on how you verified that this actually worked?

(Just for one of them - either email or sms - is sufficient)

**Review by nija-at — CHANGES_REQUESTED** [review]
Thanks for the PR @thchia.

Please see my first round of comments below.

**Comment by thchia** [comment]
@nija-at thank you for the feedback! I will update the PR.

**Comment by thchia** [comment]
Hi @corymhall. Please see my changes based on feedback by Niranjan. I'm not sure if something went wrong with CodeBuild, as the build commented that it had failed, even though it is still marked as pending in the merge checklist as I write this.

The failure looks like a git checkout error, not sure if it can be re-run?

**Review by corymhall — APPROVED** [review]
Thanks for the contribution!

**Pull request merged**